mypy/typeanal.py

```diff
@@ -562,6 +588,75 @@ def analyze_callable_args(self, arglist: TypeList) -> Optional[Tuple[List[Type],
         check_arg_kinds(kinds, [arglist] * len(args), self.fail)
         return args, kinds, names
 
+    def analyze_literal_type(self, t: UnboundType) -> Type:
+        if len(t.args) == 0:
+            self.fail('Literal[...] must have at least one parameter', t)
+            return AnyType(TypeOfAny.from_error)
+
+        output = []  # type: List[Type]
+        for i, arg in enumerate(t.args):
+            analyzed_types = self.analyze_literal_param(i + 1, arg, t)
+            if analyzed_types is None:
+                return AnyType(TypeOfAny.from_error)
+            else:
+                output.extend(analyzed_types)
+        return UnionType.make_union(output, line=t.line)
+
+    def analyze_literal_param(self, idx: int, arg: Type, ctx: Context) -> Optional[List[Type]]:
+        # This UnboundType was originally defined as a string.
+        if isinstance(arg, UnboundType) and arg.original_str_expr is not None:
+            return [LiteralType(
+                value=arg.original_str_expr,
+                fallback=self.named_type('builtins.str'),
+                line=arg.line,
+                column=arg.column,
+            )]
+
+        # If arg is an UnboundType that was *not* originally defined as
+        # a string, try expanding it in case it's a type alias or something.
+        if isinstance(arg, UnboundType):
+            arg = self.anal_type(arg)
+
+        # Literal[...] cannot contain Any. Give up and add an error message
+        # (if we haven't already).
+        if isinstance(arg, AnyType):
+            # Note: We can encounter Literals containing 'Any' under three circumstances:
+            # if the user attempts use an explicit Any as a parameter, if the user
+            # is trying to use an enum value imported from a module with no type hints,
+            # giving it an an implicit type of 'Any', or if there's some other underlying
+            # problem with the parameter.
```

Comment:
I turned this comment into a list. I think I want to avoid noting which TypeOfAny corresponds to which case for now though -- I haven't actually investigated deeply which kind corresponds to which case and want to avoid documenting false info.

Probably I'll flesh this comment out more once I circle back around to handling enums and do the thing you suggest then. I left a TODO about this.